Taxes Report: fix comparison mode and segmentation labels

psealock opened the pull request:
Fixes https://github.com/woocommerce/wc-admin/issues/1554

When a user selects "Comparison" in the Taxes Report, change the chart mode to `item-comparison` and include `?segmentby=tax_rate_id` in the API request. Also, construct the labels correctly in the segments portion of the response. 

### Detailed test instructions:

1. Taxes Report.
2. See the chart appear as `time-comparison` for two periods (selected, previous).
2. Show > Comparison.
3. See the chart in `item-comparison` mode.
4. Check that labels in the chart match those found on the table.

psealock committed:
Taxes Report: allow segmentation by tax

justinshreve commented on includes/class-wc-admin-reports-segmenting.php:
```diff
@@ -363,15 +363,25 @@ protected function set_all_segments() {
-				"SELECT tax_rate_id FROM {$wpdb->prefix}woocommerce_tax_rates
-						UNION 
-						SELECT DISTINCT meta_value FROM {$wpdb->prefix}woocommerce_order_itemmeta where meta_key='rate_id'",
+			$taxes = $wpdb->get_results(
```
I think we should consider refactoring the two queries still present in `set_all_segments`. This and coupons above use direct queries. We should instead use a data store method to make swapping things out easier. (This can be a separate `@todo`/issue/PR).

(I think even `get_categories` should be updated, since CPT access in core is also behind a data store too). We want to be using filterable functions like `wc_get_products` or data store methods and not directly touching the database or post types.

justinshreve commented on client/analytics/report/taxes/index.js:
```diff
@@ -20,22 +21,46 @@ import ReportSummary from 'analytics/components/report-summary';
 export default class TaxesReport extends Component {
+	getChartMeta() {
+		const { query } = this.props;
+		const isCompareTaxView = 'compare-tax-codes' === query.filter;
```
I think we should check `query.taxes.split( ',' ).length` here too. Otherwise, the mode immediately switches to compare mode when you change the filter, instead of after clicking the 'compare' button.

emdashcodes reviewed (changes requested):
Edit:

It might also be worth you rebasing this branch and then reprocessing your data (since https://github.com/woocommerce/wc-admin/pull/1515 merged).

--

I'm having a couple issues testing this.

1) Compare mode kicks in as soon as you switch to the 'compare' filter. On our other reports, you must select the elements to compare first. See my comment about checking the query.

2) I selected two valid codes in the dropdown, but something weird seems to be happening where one of the labels in the compare is wrong:

All view, showing I have data:

<img width="1246" alt="screen shot 2019-02-18 at 3 36 09 pm" src="https://user-images.githubusercontent.com/689165/52975338-18c0ac00-3393-11e9-949b-640be514a6f4.png">

Before compare:

<img width="1251" alt="screen shot 2019-02-18 at 3 36 25 pm" src="https://user-images.githubusercontent.com/689165/52975352-24ac6e00-3393-11e9-9b78-ea6f181a981c.png">

After hitting compare (notice the one label switched, and the one value is 0):

<img width="1343" alt="screen shot 2019-02-18 at 3 36 48 pm" src="https://user-images.githubusercontent.com/689165/52975376-2f670300-3393-11e9-89c6-5cd1888bd2d6.png">

psealock commented:
Thanks for the review @justinshreve 

RE: Removing direct queries - [WC_Tax](https://docs.woocommerce.com/wc-apidocs/class-WC_Tax.html) doesn't have a direct way of gathering all tax codes. What do you think is the best way to proceed? Add a method to `WC_Tax`? Or create a data store method in `WC_Admin_Reports_Taxes_Stats_Data_Store`?

>Compare mode kicks in as soon as you switch to the 'compare' filter. On our other reports, you must select the elements to compare first.

hmm... switching to the 'compare' filter directly produces a compare mode in Products, Variations, and Categories. Coupons does not work this way, but `getChartMeta` has yet to be implemented in that report. Am I looking at the right thing?

I do now see an issue where selecting two taxes from the chart and clicking "Compare" is not working as expected.

>but something weird seems to be happening where one of the labels in the compare is wrong

Nice find. I can reproduce and will dig into why.

justinshreve commented:
> hmm... switching to the 'compare' filter directly produces a compare mode... m I looking at the right thing?

Doh! I blame it being Monday. I even recently worked on those compare modes 😆. I just double checked and you're right -- so ignore me about that!

justinshreve commented:
> Or create a data store method in `WC_Admin_Reports_Taxes_Stats_Data_Store`

This sounds good to me.

psealock commented:
@justinshreve could you take another look?

I updated `getLabels` for taxes because the API doesn't support an `includes` parameter and simply returns an array of all taxes. Because of this I couldn't use the `getRequestByIdString` helper function.

I also created a data store method in `WC_Admin_Reports_Taxes_Stats_Data_Store` to query the database. I also simplified the query and was able to remove a `JOIN`.

emdashcodes reviewed (commented):
I'm still having an issue here where a rate shows up in the chart even though its not in the compare. The table looks fine though.

justinshreve commented on client/analytics/report/taxes/config.js:
```diff
@@ -52,15 +58,29 @@ export const filters = [
+							return Promise.resolve( [] );
+						}
+						const payload = stringifyQuery( {
+							per_page: -1,
```
I don't know if `per_page` works with -1. The API does validation and sets this between 1-100:

https://github.com/woocommerce/woocommerce/blob/d0491072e8a7d0a4535db0ec2bae4a6d43ff64fd/includes/api/v1/class-wc-rest-taxes-controller.php#L665-L673.

Long term, I think we should probably update the API to have an `include` param. That would be consistent with the other filters. That would be a bit more work on top of this PR though. Maybe this is OK for now / initial release of the feature plugin (capped at 100 tax rates via comparison which seams reasonable since you can only chart 5 at once anyway). 

tl;dr I'm OK with maybe just adding a TODO for now about adding an `include` param and the limit here and going with this.

joshuatf replied:
@justinshreve I think `apiFetch` handles `-1` in batches so this should be okay.  https://github.com/WordPress/gutenberg/blob/1020543c18a7e303e652ba5b27c4e434642ec72c/packages/api-fetch/src/middlewares/fetch-all-middleware.js#L39

But I do agree that long-term it's a good call to have an `include` param instead of fetching all.  

@psealock I think we can handle that in a separate PR if you'd like and I can reword this issue to address the include param - https://github.com/woocommerce/wc-admin/issues/1640

joshuatf replied:
I have an open PR that will add the `include` param for use here - https://github.com/woocommerce/wc-admin/pull/1672

justinshreve commented on includes/class-wc-admin-reports-segmenting.php:
```diff
@@ -370,15 +371,13 @@ protected function set_all_segments() {
-				ARRAY_A
-			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
-			$segments     = wp_list_pluck( $tax_rate_ids, 'tax_rate_id' );
+			$taxes = WC_Admin_Reports_Taxes_Stats_Data_Store::get_all_taxes();
```
I just implemented something similar here too:
https://github.com/woocommerce/wc-admin/pull/1647/

joshuatf reviewed (commented):
Nice work on this @psealock!  Good to see the tax labels working.

Left a comment about handling the `include` param in a separate issue, but this seems to be testing well for me.

Probably good to go after addressing @justinshreve's question about filtering returned taxes in segmentation.

psealock commented:
Thanks for the feedback @joshuatf @justinshreve !

I added an additional argument for filtering by specific tax rates in the data store similar to the one done for Coupons.

I also added a TODO item in the `getLabels` function about the `include` collection param and updated the title in https://github.com/woocommerce/wc-admin/issues/1640 to reflect that.

emdashcodes reviewed (approved):
Thanks for all the updates here! This is looking good and testing well for me now. 🚢 it.
